### PR TITLE
Person history

### LINKED ISSUE #91: Add affiliation history for Doctors

### LINKED ISSUE #92: Add affiliation history for Patients

### PR DESCRIPTION
**Pull request by shuenj**
Add affiliationHistory feature to Person class

Main changes:
- AffiliationHistory as an attribute in Person (if not specified, it'll use affiliations as history)
- Update addCommand to add to history when affiliation is stated
- Update EditCommand to add to history when affiliation is changed
- Test cases to include withAffiliationHistory
- Misc like adding commas to affiliations in PersonCard

For discussion:
- make Person abstract class
- replace Person in testing
- editing affiliations and affiliationHistory
- **Make changing roles illegal (to preserve affiliation type constraint)**

Future updates/for discussion:
- update affiliationHistory when name changes

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 1]
```diff
@@ -86,8 +86,10 @@ private static Person createEditedPerson(Person personToEdit, EditPersonDescript
         Role updatedRole = editPersonDescriptor.getRole().orElse(personToEdit.getRole());
         Set<Affiliation> updatedAffiliations = editPersonDescriptor
                 .getAffiliations().orElse(personToEdit.getAffiliations());
-
-        return updatedRole.generatePerson(updatedName, updatedPhone, updatedEmail, updatedAffiliations);
+        Set<Affiliation> mergedAffiliationHistory = new HashSet<>(personToEdit.getAffiliationHistory());
+        mergedAffiliationHistory.addAll(updatedAffiliations);
+        return updatedRole.generatePerson(updatedName, updatedPhone, updatedEmail,
+            updatedAffiliations, mergedAffiliationHistory);
```
Comment (leezhanpeng) [inline src/main/java/seedu/address/logic/commands/EditCommand.java:92]: Could there potentially be a bug here? While `Set` does not store duplicates, will it be able to distinguish between two different affiliations even though they may contain the same `Person` string?

If not, there may be issues of duplicate `Affiliation` in the `AffiliationHistory` when we try to merge old and new `Affiliation` sets if they both contain the same `Affiliation` but of different instance.

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 2]
```diff
@@ -255,6 +266,40 @@ public void setAffiliations(Set<Affiliation> affiliations) {
...
+        }
+        /**
+         * Adds {@code affiliations} to this object's {@code affiliations}.
+         * @param affiliations
+         */
+        public void addAffiliationsToHistory(Set<Affiliation> affiliations) {
+            if (this.affiliationHistory == null) {
+                this.affiliationHistory = new HashSet<>();
+            }
+            this.affiliationHistory.addAll(affiliations);
```
Comment (leezhanpeng) [inline src/main/java/seedu/address/logic/commands/EditCommand.java:301]: Same for this regarding duplicate affiliations even though `HashSet` is used.

File: src/test/java/seedu/address/logic/commands/EditPersonDescriptorTest.java [hunk 3]
```diff
@@ -53,10 +57,52 @@ public void equals() {
...
+        Set<Affiliation> newAffiliations = new HashSet<>();
+        newAffiliations.add(affiliation3);
+        newAffiliations.add(affiliation4);
+        descriptor.addAffiliationsToHistory(newAffiliations);
+        assertTrue(descriptor.getAffiliationHistory().isPresent());
+        Set<Affiliation> expectedHistory = new HashSet<>(initialHistory);
+        expectedHistory.addAll(newAffiliations);
+        assertEquals(expectedHistory, descriptor.getAffiliationHistory().get());
+    }
+
```
Comment (leezhanpeng) [inline src/test/java/seedu/address/logic/commands/EditPersonDescriptorTest.java:105]: Trust

File: src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java [hunk 4]
```diff
@@ -81,6 +82,17 @@ public EditPersonDescriptorBuilder withAffiliations(String... tags) {
         return this;
     }
 
+    /**
+     * Parses the {@code affiliationHistory} into a {@code Set<Affiliation>} and
+     * set it to the {@code EditPersonDescriptor}
+     * @return EditPersonDescriptorBuilder
+     */
+    public EditPersonDescriptorBuilder withAffiliationHistory(String... tags) {
```
Comment (leezhanpeng) [inline src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java]: Maybe using `affiliations` as the parameter name is better here. Likewise for the function above (javadoc as well)! Might have been missed out in previous PRs.

**Review by leezhanpeng — COMMENTED** [review]
Amazing stuff here. Just a minor concern that I've commented on as I skimmed through the code.

Apart from your indicated future update on name-changing interactions, I think it will also be good to discuss if removing a `Person` should also trigger updates in the `AffiliationHistory` records of other `Person`s. Will leave this for Tuesday.

File: src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java (hunk 4 above)
Comment (shuenj) [reply src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java]: good catch, thanks!

File: src/test/java/seedu/address/logic/commands/EditPersonDescriptorTest.java (hunk 3 above)
Comment (shuenj) [reply src/test/java/seedu/address/logic/commands/EditPersonDescriptorTest.java:105]: same tbh wgt

File: src/main/java/seedu/address/logic/commands/EditCommand.java (hunk 1 above)
Comment (shuenj) [reply src/main/java/seedu/address/logic/commands/EditCommand.java:92]: I think it relies on the Affiliation class equals() and hashCode(), but we can discuss more tomorrow with @Seonlo99

**Review by leezhanpeng — APPROVED** [review]
LGTM

File: src/main/java/seedu/address/logic/commands/EditCommand.java (hunk 1 above)
Comment (Seonlo99) [reply src/main/java/seedu/address/logic/commands/EditCommand.java:92]: Yea I think we can discuss today, we can do a check before adding the affiliation to the affiliation set.

**Pull request merged**